### ISSUE TITLE
move metaframework APIs into paraglide js

### ISSUE THREAD

**Issue by samuelstroschein** [opral/inlang-paraglide-js#185]
## Context

 I find it hard to give recommendations/support on paraglide for different metaframeowkrs because they have different APIs.

![CleanShot 2024-08-09 at 11 35 30@2x](https://github.com/user-attachments/assets/cdcc60b3-6d9b-4e13-bb3f-ba912a5c42d6)

## Proposal 

Move metaframework-related APIs like routing strategy into paraglide-js for a consistent way/knowledge transfer between metaframework.

-  Paraglide JS - defines the APIs for plain JS, frameworks, meta frameworks
-  Paraglide {next|svelte|x} - implements the APIs  I find it hard giving recommendations/support

## Additional information 

Came up in https://discord.com/channels/897438559458430986/1271469453149732995/1271491605077037097

**Comment by NilsJacobsen** [comment]
Curious what @loris.sigrist thinks about it

**Comment by LorisSigrist** [comment]
I try to keep the meta-framework APIs similar by reusing concepts that work in one in as many as I can. The `RoutingStrategy` interface is really promising & portable for example. 

However, basically everything needs to interact with framework code in some way. Even `RoutingStrategy` relies on being able to access the Framework's `Request` object. 

IMO most of the framework code would be quite difficult to centralize, however, we can certainly align the concepts more. Paraglide-SvelteKit should have a `RoutingStrategy` interface as well

**Comment by samuelstroschein** [comment]
> @LorisSigrist  However, basically everything needs to interact with framework code in some way. 

Implementation would be up to the adapters. For knowledge transfer, the only important part is the API. Agree/disagree?

**Comment by LorisSigrist** [comment]
I can try to come up with a unified API that works in all our adapters, however, I'm not sure it'll end up being simpler, as the adapters will likely end up with way more primitives to implement.

TBH I would need to try before I can give a confident assessment here

**Comment by samuelstroschein** [comment]
Makes sense. Let's delay the discussion und push out importer/exporter + variants

**Comment by samuelstroschein** [comment]
For future discussions: 

**Moving the vite plugin to paraglide js also seems to make sense. **

* In [https://github.com/orgs/opral/discussions/3056](https://github.com/orgs/opral/discussions/3056) I had to explain that you can use "paraglide vite" to use paraglide js with vite. That is confusing. Easier would be "import the vite plugin from paraglide js"
* I saw multiple users using both `paraglide-vite`, and `paraglide-sveltekit`  in the dependencies. Prob because they were confused. 

**Maybe an "adapter" architecture was/is better for communication and ease of use**

With the spin that everyone installs `paraglide-js` and an adapter "hooks" into the compiler instead of providing their own compiler. 

```
paraglide-js compile --adapter @inlang/paraglide-js-sveltekit-adapter
```

```
import defineConfig from "vite"
import { vitePlugin as paraglide } from "@inlang/paraglide-js"
import svelteKitAdapter from "@inlang/paraglide-js-sveltekit-adapter"

export default defineConfig({
  plugins: [
    paraglide({ adapter: svelteKitAdapter })
  ]
})
```

**Comment by samuelstroschein** [comment]
(No action required, just adding context)<br>Another reason for "going back to adapters" is branding. Users are referring to Paraglide JS as Paraglide JS. I found myself being confused too when I give recommendations if I should say "use Paraglide Svelte|React|Next|SvelteKit|Astro|X" instead of just "use Paraglide JS". 

*Screenshot is taken from *[*https://www.reddit.com/r/nextjs/comments/1etryqn/i18n_with_app_router/*](https://www.reddit.com/r/nextjs/comments/1etryqn/i18n_with_app_router/)

<img src="https://uploads.linear.app/4cb6d418-86ee-4425-b5c8-f91266b0ab28/605986f4-2784-4738-b23d-87090e808373/4894321b-debd-4653-8f03-dff4a3f50ce8?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzRjYjZkNDE4LTg2ZWUtNDQyNS1iNWM4LWY5MTI2NmIwYWIyOC82MDU5ODZmNC0yNzg0LTQ3MzgtYjIzZC04NzA5MGU4MDgzNzMvNDg5NDMyMWItZGViZC00NjUzLThmMDMtZGZmNGEzZjUwY2U4IiwiaWF0IjoxNzIzODI4MDkzLCJleHAiOjMzMjk0Mzg4MDkzfQ.fQ2_wtXBC6Q8m9Lel8KIpSz1U7PiGBKfuf1xPhipRFE " alt="CleanShot 2024-08-16 at 13.04.07@2x.png" width="1624" height="1182" />

**Comment by samuelstroschein** [comment]
Independent upgrading of paragide js and the metaframework implementation is another reason to go back to adapters. 

Releasing paraglide 2.0 with variant supports needs to update all metaframework versions. It would be (much!) easier if Paraglide JS can be upgraded independently of the adapters.

**Comment by samuelstroschein** [comment]
Done https://github.com/opral/inlang-paraglide-js/issues/201#issuecomment-2673375348
